Code/GraphMol/test1.cpp

```diff
@@ -1574,6 +1574,84 @@ void testGithub1843() {
   BOOST_LOG(rdErrorLog) << "Finished" << std::endl;
 }
 
+void testHasValenceViolation() {
+  BOOST_LOG(rdInfoLog) << " ----------> Testing Atom::hasValenceViolation()"
+                       << std::endl;
+  auto to_mol = [](const auto &smiles) {
+    int debugParse = 0;
+    bool sanitize = false;
+    auto mol = RDKit::SmilesToMol(smiles, debugParse, sanitize);
+    TEST_ASSERT(mol != nullptr);
+    mol->updatePropertyCache(false);
+    return mol;
+  };
+
+  //  All valid atoms
+  for (const auto &smiles : {
+           "C",
+           "C(C)(C)(C)C",
+           "S(C)(C)(C)(C)(C)C",
+           "O(C)C",
+           "[H-]",
+           "[H+] |^2:0|",  // same as [H-]
+           "[HH]",
+           "[H-2]",  // behaves like Li, which allows any valence
```

Comment:
Is this sensible?

Same question below about allowing:
```
"[H][Na]([H])([H])[H]",  // Na allows any valence
```

Reply:
H2- is definitely not stable, but the RDKit doesn't currently complain about it.
I think that's actually wrong and that it should give the same error as O3- or F2-:
```
In [6]: Chem.MolFromSmiles('[O-3]')
[17:17:43] Explicit valence for atom # 0 O, 3, is greater than permitted
```

Reply:
As for the question about Na: since we allow any valence on it, then that four-coordinate form is indeed allowed. One can argue about it.

Reply:
H-2 is now disallowed, and added tests to also make sure O-3 and F-2 are disallowed.

To make that happen, Na is now restricted to require a valence of 1. The allowance of group I/II metals is someone inconsistent when -1 is present, so I updated the "allow any" rule to be only when the allowed valence is *only* -1.